### ISSUE TITLE
How to maintain ordering guarantees during parallel execution?

### ISSUE THREAD

**Issue by sth1997** [ldbc/ldbc_snb_interactive_driver#108]
Hi.
I'm going to use interactive workload against some database. Due to some technical problems, I need some information about dependent operations. For example, is there a dependency between the update and the query？ That is, will the data be queried after being updated?

I noticed that the Introduction of wiki said that

> It is capable of generating parallel workloads (e.g. concurrent reads and writes), while respecting the configured operation mix and ensuring that ordering between dependent operations is maintained (for more details see Implementing workload operations).

and

> As this is a complex component the description of it is largely out of scope here, but a more detailed discussion of its design can be found in Maintaining ordering guarantees during parallel execution.

But I didn't find any information about dependent operations in * Implementing workload operations* and the link to *Maintaining ordering guarantees during parallel execution* has failed.

Thanks.

**Comment by szarnyasg** [comment]
Good question. I don't think the dependencies are collected anywhere.
I created a matrix showing the dependencies between Interactive updates and BI complex reads:
![image](https://user-images.githubusercontent.com/1402801/62612550-e0446880-b907-11e9-87d7-ef8e8c2f7630.png)

I can do something similar for Interactive updates and Interactive complex/short reads later this week.

Section 4.2 of the [paper](https://homepages.cwi.nl/~boncz/snb-challenge/snb-sigmod.pdf) discusses how dependent operations are handled. Let me know if this helps.

**Comment by sth1997** [comment]
Thanks!
 But to be honest, I don't quite understand the meaning of the numbers in the matrix. 😅

**Comment by sth1997** [comment]
Another question, if I set thread_count = 1, threr will be only one thread for OperationHandler instances, and the beginning of the next query must be after the completion of the previous update. This will cause no error.
But if I set thread_count = 2, one UpdateOperationHandler instance update and and another QueryOperationHandler instance begin to query before the UpdateOperationHandler instance completing update, the result of the query will be wrong.
How to avoid this kind of situation? Are queries and updates between each thread independent?

**Comment by szarnyasg** [comment]
> But to be honest, I don't quite understand the meaning of the numbers in the matrix. sweat_smile

The numbers denote the intersection of type/label names between queries. E.g. if Qx has {Person, City, LIVES_IN, KNOWS} and Qy has {Person, Company, KNOWS, WORKS_AT}, the intersection is {Person, KNOWS} and the number will be 2.

**Comment by sth1997** [comment]
You said that "Clever query engines could exploit cells with 0 values.", can ldbc driver do it?
If so, even I set thread_count > 1，when a query starts to execute, the updates being executed by other threads are  independent of the query being executed by this thread. Right?

**Comment by szarnyasg** [comment]
As far as I know, the driver does not consider the content of the queries.
But other task force members know better - they will join this discussion
soon.

**Comment by jackwaudby** [comment]
Hi,

Hopefully I can clarify a few things:

The driver generates a stream of operations, which can be executed in parallel emulating multiple clients. Dividing the operations into multiple streams is not trivial because dependencies exist between operations. For example, consider the following sequence of events: Person A is added to the network (add vertex), Person B is added (add vertex), Person A and B become friends (add edge). Person A cannot become friends with Person B until Person B is added to the network, i.e we have dependent operations.

Maintaining this ordering becomes a challenge when there are parallel streams of operations, as dependent operations can be placed in different streams and there is no guarantee execution order will preserve the order between operations. The driver handles this through dependency tracking across streams. For more info on how this is implemented see: 

- [Blog post](http://ldbcouncil.org/blog/snb-driver-part-2-tracking-dependencies-between-queries)
- [Paper](https://homepages.cwi.nl/~boncz/snb-challenge/snb-sigmod.pdf)

Also, as far as I'm aware the actual dependency information between operations is only internal to the driver and not exposed

**Comment by miratepuffin** [comment]
Hi @sth1997 

Just hoping on the end of what Jack was saying, in addition to the dependencies between updates you could have dependencies between queries and updates (as you mentioned above). In  this case it would be hoped that your database isn't returning some results from a partial update which has yet to be completed and is something that the benchmark is going to help you discover. If alternatively you mean the query running before or after an update and therefore returning a different result, the both of there are 'correct' in that they returned the right result based on the state of your graph at the time of execution. 

Perhaps if you gave us a little more detail on what you are working on we could be of more help?

All the best,
Ben
